### ISSUE TITLE
ICE regression: "missing expansion backtrace"

### ISSUE THREAD

**Issue by est31** [rust-lang/rust#41776]
The following code ICE's on nightly and beta, but does not ICE on stable:

```rust
include!(line!());
```

For an example that doesn't give an immediate compile error (but later on bc it can't find the file):

```rust
include!(concat!("help-files/", file!()));
```

The ICE looks like:

```
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

thread 'rustc' panicked at 'missing expansion backtrace', /checkout/src/libcore/option.rs:794

```

**Comment by est31** [comment]
Last working nightly: `rustc 1.17.0-nightly (e1cec5d4b 2017-03-29)`
First broken nightly: `rustc 1.18.0-nightly (5c94997b6 2017-03-30)`
[list of commits in between](https://github.com/rust-lang/rust/compare/e1cec5d4bf626f151a779323e16d62fe60117086...5c94997b6b1cca3e65ecfc5ba40c033d6838019b).

**Comment by est31** [comment]
cc @jseyfried

**Comment by Mark-Simulacrum** [comment]
fe151194e9a919ab23c2bb356767a5cbfa3811e7 is the commit which broke this, PR #40597.

**Comment by nikomatsakis** [comment]
triage: P-high

Assigning to @jseyfried, hopefully they have some time to investigate!

**Comment by jseyfried** [comment]
This is closely related to #40649; fixed in #42006.
